src/app/rpmostree-compose-builtin-tree.c

```diff
@@ -896,101 +896,34 @@ impl_commit_tree (RpmOstreeTreeComposeContext *self,
-
-      g_print ("Content Bytes Written: %" G_GUINT64_FORMAT "\n", stats.content_bytes_written);
-      g_variant_builder_add (&composemeta_builder, "{sv}", "ostree-content-bytes-written",
-                             g_variant_new_uint64 (stats.content_bytes_written));
+      statsp = &stats;
     }
-  g_print ("Wrote commit: %s\n", new_revision);
-  g_variant_builder_add (&composemeta_builder, "{sv}", "ostree-commit", g_variant_new_string (new_revision));
-  /* Since JavaScript doesn't have 64 bit integers and hence neither does JSON,
-   * store this as a string:
-   * https://stackoverflow.com/questions/10286204/the-right-json-date-format
-   * */
-  { guint64 commit_ts = ostree_commit_get_timestamp (new_commit);
-    g_autofree char *commit_ts_iso_8601 = rpmostree_timestamp_str_from_unix_utc (commit_ts);
-    g_variant_builder_add (&composemeta_builder, "{sv}", "ostree-timestamp", g_variant_new_string (commit_ts_iso_8601));
-  }
-  const char *commit_version = NULL;
-  (void)g_variant_lookup (new_commit_inline_meta, OSTREE_COMMIT_META_KEY_VERSION, "&s", &commit_version);
-  if (commit_version)
-    g_variant_builder_add (&composemeta_builder, "{sv}", "ostree-version", g_variant_new_string (commit_version));
-
-  const char *inputhash = NULL;
-  (void)g_variant_lookup (new_commit_inline_meta, "rpmostree.inputhash", "&s", &inputhash);
-  /* We may not have the inputhash in the split-up installroot case */
-  if (inputhash)
-    g_variant_builder_add (&composemeta_builder, "{sv}", "rpm-ostree-inputhash", g_variant_new_string (inputhash));
-  if (parent_revision)
-    g_variant_builder_add (&composemeta_builder, "{sv}", "ostree-parent-commit", g_variant_new_string (parent_revision));
 
-  if (opt_write_commitid_to)
-    {
-      if (!g_file_set_contents (opt_write_commitid_to, new_revision, -1, error))
-        return glnx_prefix_error (error, "While writing to '%s'", opt_write_commitid_to);
-    }
-  else if (self->ref)
+  g_print ("Wrote commit: %s\n", new_revision);
+  if (self->ref)
     {
       g_print ("%s => %s\n", self->ref, new_revision);
       g_variant_builder_add (&composemeta_builder, "{sv}", "ref", g_variant_new_string (self->ref));
     }
 
-  if (opt_write_composejson_to && parent_revision)
-    {
-      g_autoptr(GVariant) diffv = NULL;
-      if (!rpm_ostree_db_diff_variant (self->repo, parent_revision, new_revision,
-                                       FALSE, &diffv, cancellable, error))
-        return FALSE;
-      g_variant_builder_add (&composemeta_builder, "{sv}", "pkgdiff", diffv);
-    }
+  if (!rpmostree_composeutil_write_composejson (self->repo,
+                                                opt_write_composejson_to, statsp,
+                                                new_revision, new_commit,
+                                                &composemeta_builder,
+                                                error))
+    return FALSE;
 
-  if (opt_write_composejson_to)
+  if (opt_write_commitid_to)
```

Comment:
Hmm, I think this slightly changes the semantics of `--write-commitid-to` though. Before it would override `ref`, but now we're doing both. That's likely to cause issues in e.g. pungi, which does the ref update as a separate step.

Reply:
We have a test for that:
```
# --write-commitid-to should not set the ref
if ostree --repo=${repobuild} rev-parse ${treeref}; then
    fatal "Found ${treeref} ?"
fi
echo "ok ref not written"
```

which is passing.  I think the diff is just misleading.  If you look at the logic above, we're still doing:

```
  /* --write-commitid-to overrides writing the ref */
  if (self->ref && !opt_write_commitid_to)
    {
      if (use_txn)
        ostree_repo_transaction_set_ref (self->repo, NULL, self->ref, new_revision);
      else
        {
          if (!ostree_repo_set_ref_immediate (self->repo, NULL, self->ref, new_revision,
                                              cancellable, error))
            return FALSE;
        }
    }

```

Reply:
Yep, updated :arrow_down: